services/device-registry-base/src/main/java/org/eclipse/hono/deviceregistry/service/device/AbstractRegistrationService.java

```diff
@@ -359,16 +367,24 @@ public final void setTenantInformationService(final TenantInformationService ten
 
                                     LOG.debug("auto-provisioning device {} for gateway {}", deviceId, gatewayId);
                                     return addEdgeDevice(tenantId, Optional.ofNullable(deviceId), device, span)
-                                            .compose(addDeviceOk -> sendEmptyEvent(tenantId, deviceId, gatewayId, span)
-                                                .compose(sendEmptyEventOk -> processAssertRegistration(DeviceKey.from(tenantKey, deviceId), span)
-                                                    .compose(deviceRegistrationData -> {
-                                                        final JsonObject deviceData = deviceRegistrationData.getPayload()
-                                                                .getJsonObject(RegistrationConstants.FIELD_DATA, new JsonObject());
-                                                        return setAutoProvisioningNotificationSent(tenantId, deviceId, deviceData, span)
-                                                            .compose(idOperationResult -> createSuccessfulRegistrationResult(tenantId, deviceId, deviceData, span));
-                                                    })
-                                                )
-                                            );
+                                            .compose(registrationResult -> {
+                                                if (registrationResult.isError()) {
+                                                    if (registrationResult.getStatus() < 400 || registrationResult.getStatus() >= 500) {
```

Comment:
`StatusCodeMapper.from(final int statusCode, final String detailMessage)` should take care of this mapping.